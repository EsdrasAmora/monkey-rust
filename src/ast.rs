use crate::token::{Identifier, Token};
use serde::Serialize;
use smol_str::SmolStr;

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Statement {
    Let {
        identifier: Identifier,
        value: Box<Expression>,
    },
    Return(Box<Expression>),
    Expression(Box<Expression>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Expression {
    Literal(Literal),
    Identifier(Identifier),
    BinaryExp(BinaryExpression),
    UnaryExpression(UnaryExpression),
    If(IfExpression),
    IndexExpression(IndexExpression),
    Function(FunctionExpression),
    Call(CallExpression),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum BinaryOperator {
    Eq,
    NotEq,
    Lt,
    Lte,
    Gt,
    Gte,
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum UnaryOperator {
    Not,
    Minus,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct UnaryExpression {
    pub value: Box<Expression>,
    pub operator: UnaryOperator,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct CallExpression {
    pub arguments: Vec<Expression>,
    pub function: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct IndexExpression {
    pub container: Box<Expression>,
    pub index: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct FunctionExpression {
    pub parameters: Vec<Identifier>,
    pub body: BlockStatement,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct IfExpression {
    pub condition: Box<Expression>,
    pub consequence: BlockStatement,
    pub alternative: Option<BlockStatement>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct BinaryExpression {
    pub operator: BinaryOperator,
    pub lhs: Box<Expression>,
    pub rhs: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Default)]
pub struct BlockStatement(pub Vec<Statement>);

impl BlockStatement {
    pub fn new(value: Vec<Statement>) -> Self {
        Self(value)
    }
}

impl Expression {
    pub fn boxed(self) -> Box<Self> {
        Box::new(self)
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Literal {
    Int(i64),
    String(SmolStr),
    Array(Vec<Expression>),
    Hash(Vec<(Expression, Expression)>),
    True,
    False,
    Nil,
}

impl Literal {
    pub fn into_exp(self) -> Box<Expression> {
        Expression::Literal(self).boxed()
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Expression::Literal(literal)
    }
}

impl Token {
    #[inline]
    pub fn binary_expression_type(&self) -> Option<BinaryOperator> {
        match self {
            Token::Plus => Some(BinaryOperator::Add),
            Token::Minus => Some(BinaryOperator::Sub),
            Token::Slash => Some(BinaryOperator::Div),
            Token::Asterisk => Some(BinaryOperator::Mul),
            Token::Eq => Some(BinaryOperator::Eq),
            Token::NotEq => Some(BinaryOperator::NotEq),
            Token::Lt => Some(BinaryOperator::Lt),
            Token::Lte => Some(BinaryOperator::Lte),
            Token::Gt => Some(BinaryOperator::Gt),
            Token::Gte => Some(BinaryOperator::Gte),
            _ => None,
        }
    }
}
