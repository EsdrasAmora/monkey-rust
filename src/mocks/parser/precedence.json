{
    "Expression": {
        "Eq": {
            "lhs": {
                "Add": {
                    "lhs": {
                        "Literal": {
                            "Int": 3
                        }
                    },
                    "rhs": {
                        "Mul": {
                            "lhs": {
                                "Literal": {
                                    "Int": 4
                                }
                            },
                            "rhs": {
                                "Literal": {
                                    "Int": 5
                                }
                            }
                        }
                    }
                }
            },
            "rhs": {
                "Add": {
                    "lhs": {
                        "Mul": {
                            "lhs": {
                                "Literal": {
                                    "Int": 3
                                }
                            },
                            "rhs": {
                                "Literal": {
                                    "Int": 1
                                }
                            }
                        }
                    },
                    "rhs": {
                        "Mul": {
                            "lhs": {
                                "Literal": {
                                    "Int": 4
                                }
                            },
                            "rhs": {
                                "Literal": {
                                    "Int": 5
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}