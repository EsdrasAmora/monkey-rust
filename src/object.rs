use std::{
    cell::RefCell,
    collections::{hash_map::Entry, HashMap},
    fmt::{self, Display},
    ops::{Deref, DerefMut},
    rc::Rc,
};

use serde::{self, Serialize};
use smol_str::SmolStr;
use thiserror::Error;

use crate::{
    ast::{BinaryOperator, BlockStatement, UnaryOperator},
    token::Identifier,
};

#[derive(Debug, Serialize, Clone, PartialEq, Hash, Eq)]
pub enum Object {
    Nil,
    Int(i64),
    Bool(bool),
    BuiltInFn(BuiltInFn),
    Array(Array),
    HashTable(HashTable),
    String(SmolStr),
    Function(Box<Function>),
    Return(Box<Object>),
}

#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
pub enum BuiltInFn {
    Len,
    First,
    Last,
    Rest,
    Push,
}

#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
pub struct Array(pub Vec<Object>);

#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
pub struct HashTable(pub HashMap<Object, Object>);

impl HashTable {
    pub fn new(storage: HashMap<Object, Object>) -> Self {
        Self(storage)
    }
}

impl std::hash::Hash for HashTable {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        todo!("Implement Hash for HashTable")
    }
}

impl From<Array> for Object {
    fn from(vec: Array) -> Self {
        Self::Array(vec)
    }
}

impl Array {
    pub fn new(storage: Vec<Object>) -> Self {
        Self(storage)
    }

    pub fn push(&mut self, obj: Object) {
        self.0.push(obj)
    }
}

impl Deref for Array {
    type Target = Vec<Object>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for Array {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl BuiltInFn {
    pub fn name(&self) -> &'static str {
        match self {
            BuiltInFn::Len => "len",
            BuiltInFn::First => "first",
            BuiltInFn::Last => "last",
            BuiltInFn::Rest => "rest",
            BuiltInFn::Push => "push",
        }
    }
}

type Result<T> = std::result::Result<T, EvalError>;

pub const TRUE: Object = Object::Bool(true);
pub const FALSE: Object = Object::Bool(false);
pub const NIL: Object = Object::Nil;
pub const ZERO: Object = Object::Int(0);
pub const EMPTY_STRING: Object = Object::String(SmolStr::new_inline(""));

#[derive(Error, Debug)]
pub enum EvalError {
    #[error("operator `{operator:?}` not supported for value {operand:?}")]
    UnaryOpError {
        operator: UnaryOperator,
        operand: String,
    },
    #[error("operator `{operator:?}` not supported between values {lhs:?} and {rhs:?}")]
    BinaryOpError {
        operator: BinaryOperator,
        lhs: String,
        rhs: String,
    },
    #[error("{target} coercion unsuported for {value:?}")]
    CoercionError { target: &'static str, value: String },
    #[error("Identifier {0} already defined")]
    IdentifierAlreadyDefined(Identifier),
}

#[derive(Serialize, Clone)]
pub struct Environment {
    //maybe use a single hashmap with the function name as key
    curr: HashMap<SmolStr, Object>,
    #[serde(skip_serializing)]
    outer: Option<SharedEnv>,
}

impl fmt::Debug for Environment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Environment")
            .field("curr", &self.curr)
            .finish()
    }
}

pub type SharedEnv = Rc<RefCell<Box<Environment>>>;

impl Environment {
    pub fn new() -> Self {
        Self {
            curr: HashMap::new(),
            outer: None,
        }
    }

    pub fn new_enclosed(outer: SharedEnv, curr: HashMap<SmolStr, Object>) -> SharedEnv {
        Rc::new(RefCell::new(Box::new(Self {
            curr,
            outer: Some(outer),
        })))
    }

    pub fn get(&self, name: &Identifier) -> Option<Object> {
        self.curr
            .get(&name.inner())
            //Clone
            .cloned()
            .or_else(|| self.outer.as_ref().and_then(|x| x.borrow().get(name)))
            .or_else(|| match name.inner().as_str() {
                "len" => Some(Object::BuiltInFn(BuiltInFn::Len)),
                "first" => Some(Object::BuiltInFn(BuiltInFn::First)),
                "last" => Some(Object::BuiltInFn(BuiltInFn::Last)),
                "rest" => Some(Object::BuiltInFn(BuiltInFn::Rest)),
                "push" => Some(Object::BuiltInFn(BuiltInFn::Push)),
                _ => None,
            })
    }

    pub fn try_insert(&mut self, ident: &Identifier, value: Object) -> Result<()> {
        match self.curr.entry(ident.inner()) {
            Entry::Occupied(_) => Err(EvalError::IdentifierAlreadyDefined(ident.clone())),
            Entry::Vacant(entry) => {
                entry.insert(value);
                Ok(())
            }
        }
    }
}

#[derive(Serialize, Clone)]
pub struct Function {
    pub parameters: Vec<Identifier>,
    pub body: BlockStatement,
    #[serde(skip_serializing)]
    pub env: SharedEnv,
}

impl PartialEq for Function {
    fn eq(&self, other: &Self) -> bool {
        self.parameters == other.parameters
            && self.body.0.as_ptr() == other.body.0.as_ptr()
            && self.env.as_ptr() == other.env.as_ptr()
    }
}

impl Eq for Function {}

impl std::hash::Hash for Function {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.parameters.hash(state);
        self.body.0.as_ptr().hash(state);
        self.env.as_ptr().hash(state);
    }
}

impl fmt::Debug for Function {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Function")
            .field("parameters", &self.parameters)
            .field("body", &self.body)
            .finish()
    }
}

impl Display for Function {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let params = self
            .parameters
            .iter()
            .map(|x| x.inner())
            .collect::<Vec<_>>()
            .join(", ");
        write!(f, "fn({}) {{todo()!}}", params)
    }
}

impl Display for Object {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Object::Nil => write!(f, "nil"),
            Object::Int(int) => write!(f, "{}", int),
            Object::Bool(bool) => write!(f, "{}", bool),
            Object::String(str) => write!(f, "{}", str),
            Object::BuiltInFn(builtin) => write!(f, "{}", builtin.name()),
            Object::Function(function) => write!(f, "{}", function),
            Object::Return(obj) => write!(f, "{}", obj),
            Object::HashTable(hash) => write!(
                f,
                "{{{}}}",
                hash.0
                    .iter()
                    .map(|(k, v)| format!("{}: {}", k, v))
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
            Object::Array(array) => write!(
                f,
                "[{}]",
                array
                    .0
                    .iter()
                    .map(|x| x.to_string())
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        }
    }
}

impl Function {
    pub fn new(parameters: Vec<Identifier>, body: BlockStatement, env: SharedEnv) -> Self {
        Self {
            parameters,
            body,
            env,
        }
    }
}

impl Object {
    pub fn as_typeof(&self) -> &'static str {
        match self {
            Object::Nil => "nil",
            Object::Int(_) => "int",
            Object::Bool(_) => "bool",
            Object::String(_) => "string",
            Object::BuiltInFn(builtin) => builtin.name(),
            Object::Function(_) => "function",
            Object::Return(_) => "return",
            Object::Array(_) => "array",
            Object::HashTable(_) => "object",
        }
    }

    pub fn into_string(self) -> Result<SmolStr> {
        Ok(match self {
            Object::String(str) => str,
            Object::Int(int) => int.to_string().into(),
            Object::Bool(bool) => bool.to_string().into(),
            other => {
                return Err(EvalError::CoercionError {
                    target: EMPTY_STRING.as_typeof(),
                    value: other.to_string(),
                })
            }
        })
    }

    pub fn into_int(self) -> Result<i64> {
        Ok(match self {
            Object::Int(int) => int,
            Object::Bool(bool) => bool.into(),
            Object::String(value) => match value.parse() {
                Ok(int) => int,
                Err(_) => Err(EvalError::CoercionError {
                    target: ZERO.as_typeof(),
                    value: Object::String(value).to_string(),
                })?,
            },
            other => {
                return Err(EvalError::CoercionError {
                    target: ZERO.as_typeof(),
                    value: other.to_string(),
                })
            }
        })
    }

    pub fn into_bool(self) -> Result<bool> {
        Ok(match self {
            Object::Nil => false,
            Object::Int(int) => int != 0,
            Object::Bool(bool) => bool,
            Object::String(str) => !str.is_empty(),
            Object::Function(_) => true,
            other => {
                return Err(EvalError::CoercionError {
                    target: TRUE.as_typeof(),
                    value: other.to_string(),
                })
            }
        })
    }

    pub fn not(self) -> Result<Object> {
        Ok((!self.into_bool()?).into())
    }

    pub fn minus(self) -> Result<Object> {
        match self {
            Object::Int(int) => Ok(Object::Int(-int)),
            operand => {
                return Err(EvalError::UnaryOpError {
                    operator: UnaryOperator::Minus,
                    operand: operand.to_string(),
                })
            }
        }
    }

    pub fn add(self, rhs: Object) -> Result<Object> {
        Ok(match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => Object::Int(lhs + rhs),
            (Object::String(lhs), Object::String(rhs)) => {
                Object::String(format!("{}{}", lhs, rhs).into())
            }
            (lhs, rhs) => {
                return Err(EvalError::BinaryOpError {
                    operator: BinaryOperator::Add,
                    lhs: lhs.to_string(),
                    rhs: rhs.to_string(),
                })
            }
        })
    }

    pub fn sub(self, rhs: Object) -> Result<Object> {
        Ok(match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => Object::Int(lhs - rhs),
            (lhs, rhs) => {
                return Err(EvalError::BinaryOpError {
                    operator: BinaryOperator::Sub,
                    lhs: lhs.to_string(),
                    rhs: rhs.to_string(),
                })
            }
        })
    }

    pub fn mul(self, rhs: Object) -> Result<Object> {
        Ok(match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => Object::Int(lhs * rhs),
            (lhs, rhs) => {
                return Err(EvalError::BinaryOpError {
                    operator: BinaryOperator::Mul,
                    lhs: lhs.to_string(),
                    rhs: rhs.to_string(),
                })
            }
        })
    }

    pub fn div(self, rhs: Object) -> Result<Object> {
        Ok(match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => Object::Int(lhs / rhs),
            (lhs, rhs) => {
                return Err(EvalError::BinaryOpError {
                    operator: BinaryOperator::Div,
                    lhs: lhs.to_string(),
                    rhs: rhs.to_string(),
                })
            }
        })
    }

    pub fn eq(self, rhs: Object) -> bool {
        match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => lhs == rhs,
            (Object::String(lhs), Object::String(rhs)) => lhs == rhs,
            (Object::Bool(lhs), Object::Bool(rhs)) => lhs == rhs,
            (Object::Nil, Object::Nil) => true,
            _ => false,
        }
    }

    pub fn not_eq(self, rhs: Object) -> bool {
        !self.eq(rhs)
    }

    pub fn lt(self, rhs: Object) -> bool {
        match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => lhs < rhs,
            _ => false,
        }
    }

    pub fn gt(self, rhs: Object) -> bool {
        match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => lhs > rhs,
            _ => false,
        }
    }

    pub fn lte(self, rhs: Object) -> bool {
        match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => lhs <= rhs,
            _ => false,
        }
    }

    pub fn gte(self, rhs: Object) -> bool {
        match (self, rhs) {
            (Object::Int(lhs), Object::Int(rhs)) => lhs >= rhs,
            _ => false,
        }
    }
}

impl From<bool> for Object {
    fn from(bool: bool) -> Self {
        if bool {
            TRUE
        } else {
            FALSE
        }
    }
}
