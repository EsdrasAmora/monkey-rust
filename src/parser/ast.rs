#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Statement {
    Let {
        identifier: String,
        value: Box<Expression>,
    },
    Return(Box<Expression>),
    Expression(Box<Expression>),
}

//TODO: create a builder for this.
#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Expression {
    Literal(Literal),
    Identifier(String),
    Oposite(Box<Expression>),
    Not(Box<Expression>),
    Eq(BynaryExpression),
    NotEq(BynaryExpression),
    LT(BynaryExpression),
    GT(BynaryExpression),
    Add(BynaryExpression),
    Sub(BynaryExpression),
    Mul(BynaryExpression),
    Div(BynaryExpression),
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct BynaryExpression {
    lhs: Box<Expression>,
    rhs: Box<Expression>,
}
//maybe precedence should be a method on Token.
impl Expression {
    fn boxed(self) -> Box<Self> {
        Box::new(self)
    }

    fn precedence(&self) -> u8 {
        match self {
            Expression::Literal(_) | Expression::Identifier(_) => 0,
            Expression::Oposite(_) | Expression::Not(_) => 1,
            Expression::Eq(_) | Expression::NotEq(_) => 2,
            Expression::LT(_) | Expression::GT(_) => 3,
            Expression::Mul(_) | Expression::Div(_) => 4,
            Expression::Add(_) | Expression::Sub(_) => 5,
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Literal {
    Int(i64),
    String(String),
    True,
    False,
    Nill,
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Expression::Literal(literal)
    }
}
