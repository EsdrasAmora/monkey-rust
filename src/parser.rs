use self::ast::Statement;
use self::token_parser::TokenParser;
use crate::ast;
use crate::lexer::Lexer;
use crate::token_parser;
use anyhow::Error;

#[derive(Debug)]
pub struct Parser {
    pub nodes: Vec<Statement>,
    pub errors: Vec<Error>,
}

impl Parser {
    pub fn new(lexer: Lexer) -> Self {
        let mut nodes = Vec::new();
        let mut errors = Vec::new();
        let mut tokens = TokenParser::new(lexer.tokens);

        while let Some(current) = tokens.next() {
            tokens
                .parse_statement(current)
                .map_or_else(|err| errors.push(err), |val| nodes.push(val))
        }

        Parser { nodes, errors }
    }
}

//TODO: missing Assign support, only able to declare variables with let but not assign them afterwards;
//TODO: create error types and think about when to buble them up
#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_yaml_snapshot;

    #[test]
    fn parse_call_expression() {
        let input = "
        myFn();
        myFn(1,2);
        myFn(1, 2 * 3, -4 + 5);
        myFn(fn() {});
        fn() {}();
        fn(x) { return x; }(1);
        -myFn(1,2);
        myFn(fn(a,b) {return a + b;}(1,2));";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_fn_expression() {
        let input = "
        fn() {};
        fn(x) { x; };
        fn(x, y, z) { return x + y + z; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_grouped_expression() {
        let input = "(5 + 5) * 5; -(5 + 5);";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);

        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_if_expression() {
        let input = "if (x < y) { return x; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_if_else_expression() {
        let input = "
        if (x < y) { x; } else { y; };
        if (x < y) { return x; } else { return y; };
        if (x < y) { x == 3; } else { x != 1; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_infix_expression() {
        let input = "
        5 + 5;
        5 - 5;
        5 * 5;
        5 / 5;
        5 > 5;
        5 < 5;
        5 == 5;
        5 != 5;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);

        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_prefix_expression() {
        let input = "
        -1;
        !5;
        !!-2;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_expression_statement_identifier() {
        let input = "foobar;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_expression_statement_integer_literal() {
        let input = "3;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_let_statement() {
        let input = "
        let x = false;
        let y = 10;
        let foobar = true;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_return_statement() {
        let input = "
        return 5;
        return 10;
        return 993322;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn check_parser_precedence() {
        let input = r#"
        -a * b;
        !-a;
        a + b + c;
        a + b - c;
        a * b * c;
        a * b / c;
        a + b / c;
        a + b * c + d / e - f;
        3 + 4; -5 * 5;
        5 > 4 == 3 < 4;
        5 < 4 != 3 > 4;
        3 + 4 * 5 == 3 * 1 + 4 * 5;
        true;
        false;
        3 > 5 == false;
        3 < 5 == true;
        1 + (2 + 3) + 4;
        (5 + 5) * 2;
        2 / (5 + 5);
        (5 + 5) * 2 * (5 + 5);
        -(5 + 5);
        !(true == true);
        a + add(b * c) + d;
        add(a, b, 1, 2 * 3, 4 + 5, add(6, 7 * 8));
        add(a + b + c * d / f + g);"#;
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }
}
