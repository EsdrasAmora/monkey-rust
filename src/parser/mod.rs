mod ast;

use std::default;
use std::iter::Peekable;

//TODO: how to rexport this?
use crate::lexer::token::Token;
use crate::lexer::Lexer;
use ast::Node;
use id_arena::Arena;

use self::ast::Literal;

struct Program {
    nodes: Arena<Node>,
}

impl Program {
    fn new(lexer: Lexer) -> Self {
        let mut nodes = Arena::with_capacity(32);
        let mut tokens = lexer.tokens.into_iter();

        Program { nodes }
    }

    fn new_helper(
        current: &Token,
        arena: &mut Arena<Node>,
        tokens: &mut Peekable<impl Iterator<Item = Token>>,
        //return an result
    ) -> Option<Node> {
        match current {
            Token::Let => {
                //currently does not autoformat lmao: https://github.com/rust-lang/rustfmt/issues/4914
                let Some(Token::Identifier(name)) = tokens.peek().cloned() else { return None; };
                tokens.next();

                let Some(Token::Assign) = tokens.peek() else {
                    return None;
                };
                tokens.next();

                while tokens.next().filter(|x| x != &Token::Semicolon).is_some() {}

                //TODO: remove clone
                let let_statment = Node::Let {
                    name,
                    value: arena.alloc(Node::Literal(Literal::Int(1))),
                };
                Some(let_statment)
            }
            _ => None,
        }
    }
}
