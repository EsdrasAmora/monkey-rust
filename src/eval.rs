use crate::ast::{BinaryExpression, BlockStatement, Expression, IfExpression, Literal, Statement};
use crate::object::{Environment, Object, NIL};
use crate::parser::Parser;
use anyhow::Result;

impl Environment {
    pub fn eval_program(&mut self, parser: Parser) -> Result<Object> {
        let mut result = Object::Nil;
        for node in parser.nodes {
            result = node.eval_statement()?;
            if let Object::Return(inner) = result {
                return Ok(*inner);
            }
        }
        Ok(result)
    }
}

impl Statement {
    fn eval_statement(self: Statement) -> Result<Object> {
        match self {
            Statement::Let { identifier, value } => todo!(),
            Statement::Return(exp) => Ok(Object::Return(Box::new(exp.eval()?))),
            Statement::Expression(exp) => Ok(exp.eval()?),
        }
    }
}

impl BlockStatement {
    fn eval_block(self) -> Result<Object> {
        let mut result = Object::Nil;
        for node in self.0 {
            result = node.eval_statement()?;
            if let Object::Return(_) = result {
                break;
            }
        }
        Ok(result)
    }
}

impl Expression {
    fn eval(self) -> Result<Object> {
        Ok(match self {
            Expression::Literal(literal) => match literal {
                Literal::Int(integer) => Object::Int(integer),
                Literal::True => Object::Bool(true),
                Literal::False => Object::Bool(false),
                Literal::String(string) => Object::String(string),
                Literal::Nil => Object::Nil,
            },
            Expression::Identifier(ident) => todo!(),
            Expression::Oposite(exp) => exp.0.eval()?.oposite()?,
            Expression::Not(exp) => exp.0.eval()?.not()?,
            Expression::Eq(exp) => exp.lhs.eval()?.eq(exp.rhs.eval()?).into(),
            Expression::NotEq(exp) => exp.lhs.eval()?.not_eq(exp.rhs.eval()?).into(),
            Expression::Lt(exp) => exp.lhs.eval()?.lt(exp.rhs.eval()?).into(),
            Expression::Lte(exp) => exp.lhs.eval()?.lte(exp.rhs.eval()?).into(),
            Expression::Gt(exp) => exp.lhs.eval()?.gt(exp.rhs.eval()?).into(),
            Expression::Gte(exp) => exp.lhs.eval()?.gte(exp.rhs.eval()?).into(),
            Expression::Add(exp) => exp.lhs.eval()?.add(exp.rhs.eval()?)?,
            Expression::Sub(exp) => exp.lhs.eval()?.sub(exp.rhs.eval()?)?,
            Expression::Mul(exp) => exp.lhs.eval()?.mul(exp.rhs.eval()?)?,
            Expression::Div(exp) => exp.lhs.eval()?.div(exp.rhs.eval()?)?,
            Expression::If(exp) => exp.eval()?,
            Expression::Function(_) => todo!(),
            Expression::Call(_) => todo!(),
        })
    }
}

impl BinaryExpression {
    fn eval(self) -> Result<(Object, Object)> {
        Ok((self.lhs.eval()?, self.rhs.eval()?))
    }
}

impl IfExpression {
    //, environment: &mut Environment
    fn eval(self) -> Result<Object> {
        let condition = self.condition.eval()?.into_bool()?;

        Ok(if condition {
            self.consequence.eval_block()?
        } else if let Some(alternative) = self.alternative {
            alternative.eval_block()?
        } else {
            NIL
        })
    }
}

#[cfg(test)]
mod tests {

    use super::*;
    use crate::{lexer::Lexer, parser::Parser};
    use insta::assert_yaml_snapshot;

    fn parse_program(input: &str) -> Object {
        let lexer = Lexer::new(input);
        let parser = Parser::new(lexer);
        let mut environment = Environment::new();

        environment.eval_program(parser).unwrap()
    }

    fn parse_test_input(input: &[&str]) -> Vec<Object> {
        input.iter().map(|x| parse_program(x)).collect()
    }

    #[test]
    fn eval_numeric_expression() {
        let input = [
            "5;",
            "-10;",
            "3 * (3 + 3) + 10;",
            "(5 + 10 * 2 + 15 / 3) * 2 + -10;",
        ];

        let result = parse_test_input(input.as_slice());
        assert_yaml_snapshot!(result);
    }

    #[test]
    fn eval_boolean_expression() {
        let input = [
            "true;",
            "false;",
            "1 >= 2;",
            "1 >= 1;",
            "1 <= 2;",
            "1 <= 1;",
            "1 > 2;",
            "1 > 1;",
            "1 < 2;",
            "1 < 1;",
            "true == true;",
            "false != false;",
        ];

        let result = parse_test_input(input.as_slice());
        assert_yaml_snapshot!(result);
    }

    #[test]
    fn eval_not_expression() {
        let input = [
            "!5;", "!!5;", "!0;", "!!0;", "!true;", "!!true;", "!false;", "!!false;",
        ];

        let result = parse_test_input(input.as_slice());
        assert_yaml_snapshot!(result);
    }

    #[test]
    fn eval_if_expression() {
        let input = [
            "if (true) { 10; };",
            "if (false) { 10; };",
            "if (1) { 10; };",
            "if (1 < 2) { 10; };",
            "if (1 > 2) { 10; };",
            "if (1 > 2) { 10; } else { 20; };",
            "if (1 < 2) { 10; } else { 20; };",
        ];

        let result = parse_test_input(input.as_slice());
        assert_yaml_snapshot!(result);
    }

    #[test]
    fn eval_return_statement() {
        let input = [
            "return 10;",
            "return 10; 9;",
            "return 2 * 5; 9;",
            "9; return 2 * 5; 9;",
            "if (true) { return 10; }; return 9;",
            "if (true) { if (true) { return 10; }; return 1;};",
        ];

        let result = parse_test_input(input.as_slice());
        assert_yaml_snapshot!(result);
    }

    #[test]
    fn eval_type_errors() {
        let input = [
            "5 + true;",
            "5 + true; 5;",
            "-true;",
            "true + false;",
            "true + false + true + false;",
            "5; true + false; 5",
            "if (10 > 1) {
                if (10 > 1) {
                    return true + false;
                };
                return 1;
            };",
        ];

        let result: Vec<String> = input
            .iter()
            .flat_map(|x| {
                let lexer = Lexer::new(x);
                let parser = Parser::new(lexer);
                let mut environment = Environment::new();
                let val = environment.eval_program(parser);
                match val {
                    Ok(val) => None,
                    Err(err) => Some(err.to_string()),
                }
            })
            .collect();

        assert_yaml_snapshot!(result);
    }
}
