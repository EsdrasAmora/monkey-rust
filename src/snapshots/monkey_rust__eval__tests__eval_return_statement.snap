---
source: src/eval.rs
expression: result
---
- Int: 10
- Int: 10
- Int: 10
- Int: 10
- Int: 10
- Int: 10

